Add a test that an odd capacity `BufEncoder` does not compile

jamillambert opened the pull request:
An even capacity `BufEncoder` is now enforced and should be tested to make sure it does not compile if an odd capacity is used.

Adds a crate in tests that uses an odd capacity `BufEncoder`.

Adds a CI test that passes if the odd capacity crate fails to compile.

Close #155

Kixunil commented on tests/compiletests.rs:
```diff
@@ -0,0 +1,13 @@
+    config.src_base = PathBuf::from("tests/compiletest");
+    config.target_rustcflags = Some("-L target/debug/deps".to_string());
+    compiletest::run_tests(&config);
+}
```
Looks quite complicated. Why not simply have a new CI step to run `if cargo check --something; then exit 1; fi`?

jamillambert replied:
I wasn't sure about adding a CI step just for this one, should be simple, test.  But it does work, and there might be reasons in the future to add to the compile tests?

apoelstra replied:
I think it's probably worthwhile to do this in a general way (modifying CI). If we had the means to add compile-tests I'll bet we'd add a lot of them.

Kixunil replied:
I prefer not to take dev dependency if the same can be done in CI.

tcharding commented:
Cool, looking good. I think this should be called `odd-capacity` instead of `compiletest` because we can only ever have a single test in it (since one cannot prove two build failures at the same time). And for that reason perhaps inline the function into `main`.

FTR I changed the 3 to a 4 and the crate builds - to check for no false positives. This crate risks going stale consider this:

1. We change the import path of `BufEncode` - now the test crate does not build because of that
2. We change the `BufEncode` type accidentally so that odd value builds - now the test does not catch it.

jamillambert commented:
I have changed it so that there are two compile checks now.  First it checks that an even capacity compiles, then it checks that the odd capacity doesn't.  The only difference between the two is changing the capacity from 4 to 3.

I also moved them to a sub directory in tests called `compiletests` so it is a bit tidier, especially if more are going to be added.

apoelstra commented:
Can you add a further subdirectory level `pass/` and `fail/` and have the script use the directory names to figure out whether this should be a pass or fail test?

jamillambert commented:
I have changed the script to be general using the two sub directories `pass/` and `fail/`.

I added the `BufEncoder` capacity evenness check as a separate commit.

tcharding commented:
Looks good, few `bash` things specific to the `rust-bitcoin` org

- We use `#!/usr/bin/env bash` so that Nix users can run the script (they don't typically have `/bin/bash` I'm told)
- Tend to use `set -euo pipefail` unless some reason not to
- Try to lint scripts with `shellcheck foo.sh` https://www.shellcheck.net/
- Tend to use `REPO_DIR=$(git rev-parse --show-toplevel)` to get at the repo directory
- If you feel like it (but its mainly only me that does it) you can use a bunch of helper functions that were originally written by the Rust project a long time ago. They are in various shell scripts around the place, `bitcoin/contrib/check-for-api-changes.rs` has them)

tcharding commented:
Can you put a `-` separator in `compiletests` directory please mate. Feel free to say 'f**k you are annoying when you review Tobin'.

Kixunil commented on contrib/compile-tests.sh:
```diff
@@ -0,0 +1,57 @@
+    for dir in pass/*; do
+        if [ -d "$dir" ]; then
+            pushd "$dir" > /dev/null
+            if ! cargo run > /dev/null 2>&1; then
```
The test says "compile" but it uses `run`. Also we should keep the output here.

Kixunil replied:
This is the `pass` branch though.

Kixunil commented on contrib/compile-tests.sh:
```diff
@@ -0,0 +1,57 @@
+    if [ "$all_passed" = true ]; then
+        return 0
+    else
+        return 1
```
Perhaps just define `result = 0` at the beginning and assign it to `1` on fail and just `return "$result"` to get rid of if.

jamillambert replied:
Thanks, this is a nicer way of doing it. I have changed it as suggested.

Kixunil commented on contrib/compile-tests.sh:
```diff
@@ -0,0 +1,57 @@
+    for dir in fail/*; do
+        if [ -d "$dir" ]; then
+            pushd "$dir" > /dev/null
+            if cargo run > /dev/null 2>&1; then
```
Also it should be `build`, not `run`.

jamillambert replied:
done

Kixunil commented on contrib/compile-tests.sh:
```diff
@@ -0,0 +1,57 @@
+    if [ "$all_failed" = true ]; then
+        return 0
+    else
+        return 1
```
Same as above.

jamillambert replied:
done

Kixunil commented on contrib/compile-tests.sh:
```diff
@@ -0,0 +1,53 @@
+    for dir in pass/*; do
+        if [ -d "$dir" ]; then
+            pushd "$dir" > /dev/null
+            output=$(cargo build 2>&1)
```
This shouldn't be needed but I guess it's OK. I don't expect the output to be too large.

Kixunil commented on contrib/compile-tests.sh:
```diff
@@ -0,0 +1,53 @@
+            pushd "$dir" > /dev/null
+            output=$(cargo build 2>&1)
+            if [ $? -eq 0 ]; then
+                echo "$output"
```
If compilation succeeded and it shouldn't there will be no useful information printed by `cargo`. I'd rather get rid of it.

jamillambert replied:
I have removed the cargo output for this case

jamillambert commented:
Looking at the script again I noticed the two functions are now basically identical.  I have refactored it to have a single function that handles both cases.

Kixunil commented on contrib/compile-tests.sh:
```diff
@@ -0,0 +1,43 @@
+            echo "Compiling $dir"
+            output=$(cargo build 2>&1)
+            if [ "$?" -ne "$expected_status" ]; then
+                if [ "$expected_status" -eq 0 ]; then
```
Why have another argument if it can be inferred from the first one and you're not comparing them anyway?

Kixunil replied:
Too many possibilities that we're not using are just complicating the code. Honestly, I think even having the `pass` option is a stretch (normal tests do that already!) but I understand it's for the symmetry since there are related test (one for 3, other 4) so I'm accepting it. I was really expecting this to be one line in the workflow file.

jamillambert replied:
@Kixunil you still have `requested changes` marked for this PR.  Sorry I may have misunderstood your previous comment, I read it as you are accepting it as it is?

Do you still require me to remove the expected status argument? in which case there is no reason to have a separate function at all, i.e. is below preferable the the current function in this PR?
```shell
for dir in "$REPO_DIR"/tests/compile-tests/pass/*/; do
    pushd "$dir" > /dev/null
    echo "Compiling $dir"
    if ! cargo build > /dev/null 2>&1; then
        echo "error: $dir/ failed to compile"
        exit 1
    fi
    popd > /dev/null
done

for dir in "$REPO_DIR"/tests/compile-tests/fail/*/; do
    pushd "$dir" > /dev/null
    echo "Compiling $dir"
    if cargo build > /dev/null 2>&1; [ $? -ne 101 ]; then
        echo "error: $dir/ compiled when it should not have"
        exit 1
    fi
    popd > /dev/null
done
```

Kixunil replied:
Hmm, I think yes, I find it more readable without function than with it. I don't think this is a SSOT issue because the two loops do quite different things and only share tiny bit of code.

But I don't *require* it. I clicked "request changes" before because of run vs build confusion.

jamillambert replied:
Thanks.  Since it's already approved I'll leave it as is, but I have the above code ready to go if needed.

Kixunil commented on contrib/compile-tests.sh:
```diff
@@ -0,0 +1,43 @@
+    local output
+    # All crates in the specified subdirectory must compile or fail based on expected_status
+    for dir in "$dir_path"/*; do
+        if [ -d "$dir" ]; then
```
I've just realized this check turns typos in the function invocation into silent passes which is not great. I'm not going to block the PR over it but would be really nice to address.

Kixunil commented on tests/compile-tests/fail/odd-capacity/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+name = "odd-capacity"
+version = "0.1.0"
+edition = "2021"
+rust-version = "1.63.0"
```
Side note: reminds me there's a proposal to have "rust scripts" - `.rs` files where `Cargo.toml` can be embedded (but not required). We could use it instead when it comes out.

Kixunil reviewed (approved):
ACK 853f493763b4a187020e1c7ed0eb7e94ef613711
